tripal/tripal_analysis_expression
# Fixes for gene_expression_data field

Issues Addressed
===========
This PR fixes the following issues:

- #398 : the field crashes the gene page
- #400 : the expression bargraph becomes too big with lots of samples (e.g. 476).
- #402:  the field does more than one thing and can't work with web services. 

Changes
=====
1.  To fix issue #402 I created a new field named `local__gene_expression_bargraph` and I moved all of the code for creating the bargraph into htat field.  I kept the download links and statistics (counts of features and samples) in the existing `data__gene_expression_data` field.  I also added `elementInfo()` functions to each of those fields so that they would work with web services and set appropriate CV terms for each of the values.  See the screenshot of the web-services JSON below.
2. The problem with issue #398 was that the `load()` function of the field was doing a ton of queries for 475 samples and caused the page to timeout.  The field is meant to provide a viewer and so doesn't need to load all of the data in the `load()` function. I took out that code from the load() function. I created a Drupal form with AJAX support that can then load the data for the viewer using an AJAX call.  This allowed the gene page to load. I also optimized a bit the code gets the data for the plots to make it faster.   
3. The changes above meant that the template files were no longer needed so I removed them.  There are no more template files in the module.  
4. To fix issue #400 I tweaked the JS code to set the bar width appropriate for the number of samples.  

Screenshots
========
Feature Page
--------------
A feature page with both of the new fields mentioned above. 
Note that:
- The image is showing all 475 samples and they fit in the graph.  
- The form is now Drupal style. Previously it was built using JavaScript and the theme didn't match.  The form elements all works are before.  
- Clicking on a bar gives information about the sample (shown in the popout)

<img width="339" alt="image" src="https://user-images.githubusercontent.com/1719352/155875577-f57ceb83-320e-4e69-9062-38a2b4e2af84.png">

Analysis Page
---------------
The summary is now a table but without the file download link. That functionality is removed in this PR.  This is because it conflicts with the Tripal file module.  The Tripal file module makes adding files to any content type much easier and provides license information.  User's can now download the expression matrix via the File pane that gets added by the File module. 

<img width="358" alt="image" src="https://user-images.githubusercontent.com/1719352/155875672-b02646c6-7aac-4b91-a5df-ff0814699742.png">

Web Services
--------------
The field is now fully compatible with web services and provides the same information on the page as it does in web services without, of course, the bargraph.

<img width="657" alt="image" src="https://user-images.githubusercontent.com/1719352/155875752-f74c77d5-6d67-4cdb-9389-b85be9b46fa7.png">


Testing
=====
Note: don't test this until PRs #397 and #392 have been merged in. This is because they have database updates and their ordering is before this one which also has an update.   There will most likely be a merge conflict that needs to be resolved before this should be tested.  

1. Pull the updates
2. Clear all the cache (`drush cc all`) 
3. Run the database updates (`drush updatedb` )
4. Go to Admin > Structure > Tripal Content Types > Gene (or mRNA) >Manage Fields. Then click the 'check for new fields' link at the top. This will add the new bargraph field.  
5. Go to the "Manage Display" link and move the new bargraph field so it's viewable on the page. I moved it just above the existing "Expression Data" field so they would appear together.  
6. Test loading a gene page and an analysis page.
